### PR TITLE
[3426] Add delete method to allocations

### PR DESCRIPTION
**Pull request by SamJamCul**
### Context

https://trello.com/c/T6kURpH0/3426-m-change-initial-allocation-requests

### Changes proposed in this pull request

* Add a delete endpoint for allocations
* Modify update method to allow for number of places change without request type
* Modify allocation model to use include feature

### Checklist

- [ ] Make sure all information from the Trello card is in here
- [ ] Attach to Trello card
- [ ] Rebased master
- [ ] Cleaned commit history
- [ ] Tested by running locally

### CODE REVIEW

**Comment by SamJamCul** [comment]
[[M] Change "initial" allocation requests.](https://trello.com/c/T6kURpH0/3426-m-change-initial-allocation-requests)

File: app/controllers/api/v2/allocations_controller.rb [hunk 1]
```diff
@@ -42,6 +44,15 @@ def update
         end
       end
 
+      def destroy
```
Comment (TobyRet) [inline app/controllers/api/v2/allocations_controller.rb:47]: 'delete' or 'destroy'? Your PR title says 'delete'

File: spec/requests/api/v2/allocations_spec.rb [hunk 2]
```diff
@@ -58,6 +58,29 @@
     end
   end
 
+  describe "DELETE" do
```
Comment (TobyRet) [inline spec/requests/api/v2/allocations_spec.rb:61]: Same here?

File: app/controllers/api/v2/allocations_controller.rb (hunk 1 above)
Comment (defong) [reply app/controllers/api/v2/allocations_controller.rb:47]: https://stackoverflow.com/questions/22757450/difference-between-destroy-and-delete

File: spec/requests/api/v2/allocations_spec.rb [hunk 3]
```diff
@@ -58,6 +58,29 @@
...
+    context "with an existing allocation" do
```
Comment (defong) [inline spec/requests/api/v2/allocations_spec.rb:62]: feels wide, is there a reason not restricting or tests covering
- [ ] user association
- [ ] recruitment cycle

File: spec/requests/api/v2/allocations_spec.rb (hunk 2 above)
Comment (SamJamCul) [reply spec/requests/api/v2/allocations_spec.rb:61]: I'm not quite sure what the right way to phrase it is - it's a DELETE method in the API call, but it's a destroy method in the allocations controller!

File: spec/requests/api/v2/allocations_spec.rb (hunk 3 above)
Comment (SamJamCul) [reply spec/requests/api/v2/allocations_spec.rb:62]: For the delete method we haven't put any scoping on recruitment cycle - it only locates the record by id. 

For the user association, I've added a test where the user isn't from that accredited body!

Comment (defong) [reply spec/requests/api/v2/allocations_spec.rb:62]: can add the scope for recruitment cycle then please... ther eis no reason to allow for deletion fo something out of bounds

File: app/controllers/api/v2/allocations_controller.rb [hunk 4]
```diff
@@ -42,8 +44,26 @@ def update
...
+        authorize @allocation = Allocation.find(params[:id])
+
+        if not_current_recruitment_cycle?
+          return render jsonapi_errors: ["not current recruitment cycle"], status: :unprocessable_entity
+        end
+
+        if @allocation.delete
```
Comment (defong) [inline app/controllers/api/v2/allocations_controller.rb:54]: maybe

``` ruby
  def safe_delete(recruitment_cycle:)
    if recruitment_cycle == self.recruitment_cycle
      self.delete
    else
      errors.add(:safe_delete, "not meet")
    end
  end
```

Comment (SamJamCul) [reply app/controllers/api/v2/allocations_controller.rb]: Done!

**Pull request merged**